File: bfs_dfs/인구이동.py
import sys
from collections import deque
import copy

input = sys.stdin.readline

dr = [1, -1, 0, 0]
dc = [0, 0, 1, -1]


def bfs(s_r, s_c, board, visited, new_board):
    total = 0
    nations = 0
    queue = deque()
    union_list = []
    queue.append((s_r, s_c))
    union_list.append((s_r, s_c))
    visited[s_r][s_c] = 1
    total += board[s_r][s_c]
    nations += 1

    while queue:
        r, c = queue.popleft()

        for i in range(4):
            nr = r + dr[i]
            nc = c + dc[i]

            if nr < 0 or nr >= N or nc < 0 or nc >= N:
                continue

            if visited[nr][nc]:
                continue

            if L <= abs(board[nr][nc] - board[r][c]) <= R:
                queue.append((nr, nc))
                visited[nr][nc] = 1
                union_list.append((nr, nc))
                total += board[nr][nc]
                nations += 1

    new_population = total // nations
    for u in union_list:
        new_board[u[0]][u[1]] = new_population


N, L, R = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(N)]

cnt = 0
while True:
    visited = [[0] * N for _ in range(N)]
    new_board = copy.deepcopy(board)
    flag = False
    for r in range(N):
        for c in range(N):
            for i in range(4):
                nr = r + dr[i]
                nc = c + dc[i]

                if visited[r][c]:
                    continue

                if nr < 0 or nr >= N or nc < 0 or nc >= N:
                    continue

                if L <= abs(board[nr][nc] - board[r][c]) <= R:
                    bfs(r, c, board, visited, new_board)
                    flag = True

    if not flag:
        break
    board = copy.deepcopy(new_board)
    cnt += 1

print(cnt)

File: bfs_dfs/인구이동.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <queue>
#include <vector>
#define VV(T) vector<vector<T>>
#define V(T) vector<T>
#define FOR(i, m) for (int i = 0; i < m; i++)
using namespace std;

int N, L, R;
int dr[4] = {1, -1, 0, 0};
int dc[4] = {0, 0, 1, -1};

void bfs(int sr, int sc, VV(int) & board, VV(int) & visited,
         VV(int) & newBoard) {
  int total = 0;
  int nations = 0;
  queue<pair<int, int>> queue;
  vector<pair<int, int>> unionVector;

  queue.push({sr, sc});
  unionVector.push_back({sr, sc});
  visited[sr][sc] = 1;
  total += board[sr][sc];
  nations++;

  while (!queue.empty()) {
    int r, c;
    r = queue.front().first;
    c = queue.front().second;
    queue.pop();

    FOR(i, 4) {
      int nr = r + dr[i];
      int nc = c + dc[i];

      if (nr < 0 || nr == N || nc < 0 || nc == N) {
        continue;
      }

      if (visited[nr][nc]) {
        continue;
      }

      int diff = abs(board[nr][nc] - board[r][c]);
      if (L <= diff & diff <= R) {
        queue.push({nr, nc});
        visited[nr][nc] = 1;
        unionVector.push_back({nr, nc});
        total += board[nr][nc];
        nations += 1;
      }
    }
  }
  int newPopulation = total / nations;
  for (auto u : unionVector) {
    newBoard[u.first][u.second] = newPopulation;
  }
}

int main() {

  cin >> N >> L >> R;

  VV(int) board(N, V(int)(N, 0));

  FOR(i, N) {
    FOR(j, N) { cin >> board[i][j]; }
  }

  int cnt = 0;
  while (true) {
    VV(int) visited(N, V(int)(N, 0));
    VV(int) newBoard = board;
    bool flag = false;

    FOR(r, N) {
      FOR(c, N) {
        FOR(i, 4) {
          int nr = r + dr[i];
          int nc = c + dc[i];

          if (visited[r][c]) {
            continue;
          }

          if (nr < 0 || nr >= N || nc < 0 || nc >= N) {
            continue;
          }

          int diff = abs(board[nr][nc] - board[r][c]);
          if (L <= diff && diff <= R) {
            bfs(r, c, board, visited, newBoard);
            flag = true;
          }
        }
      }
    }
    if (!flag) {
      break;
    }

    board = newBoard;
    cnt++;
  }
  cout << cnt;
}